packages/evm/test/core/signature-validator.js

```diff
@@ -0,0 +1,611 @@
+const abi = require('web3-eth-abi')
+const { ecsign } = require('ethereumjs-util')
+const { bn, ZERO_BYTES32 } = require('@aragon/contract-helpers-test')
+const { keccak256, toHex, padLeft, soliditySha3 } = require('web3-utils')
+const { assertAmountOfEvents, assertEvent, assertRevert, assertBn } = require('@aragon/contract-helpers-test/src/asserts')
+
+const { SIGNATURES_VALIDATOR_ERRORS } = require('../helpers/utils/errors')
+
+const SignaturesValidator = artifacts.require('SignaturesValidatorMock')
+
+contract('SignaturesValidator', ([_, sender, strange]) => {
+  let validator
+  const wallet = web3.eth.accounts.create('erc3009')
+  const externalAccount = wallet.address
+  const externalAccountPK = wallet.privateKey
+
+  const sign = message => {
+    const { v, r, s } = ecsign(Buffer.from(message.replace('0x', ''), 'hex'), Buffer.from(externalAccountPK.replace('0x', ''), 'hex'))
+    return { v, r: `0x${r.toString('hex')}`, s: `0x${s.toString('hex')}` }
+  }
+
+  const encodeSignature = (calldata, deadline, { v, r, s }) => {
+    const encodedDeadline = padLeft(toHex(deadline).slice(2), 64)
+    const encodedV = padLeft(toHex(v).slice(2), 64)
+    const encodedR = r.slice(2)
+    const encodedS = s.slice(2)
+    return `${calldata}${encodedDeadline}${encodedV}${encodedR}${encodedS}`
+  }
+
+  before('deploy validator', async () => {
+    validator = await SignaturesValidator.new()
+  })
+
+  describe('decoding', () => {
+    const calldata = abi.encodeFunctionSignature('decodeCalldata()')
+
+    context('when there is no signature encoded', () => {
+      it('decodes empty data', async () => {
+        const receipt = await validator.sendTransaction({ data: calldata, from: sender })
+
+        assertAmountOfEvents(receipt, 'CalldataDecoded')
+        assertEvent(receipt, 'CalldataDecoded', { expectedArgs: { data: calldata, deadline: 0, v: 0, r: ZERO_BYTES32, s: ZERO_BYTES32 } })
+      })
+    })
+
+    context('when there is a signature encoded', () => {
+      const deadline = 15
+
+      it('decodes it properly', async () => {
+        const signature = sign(soliditySha3('message'))
+        const calldataWithSignature = encodeSignature(calldata, deadline, signature)
+
+        const receipt = await validator.sendTransaction({ data: calldataWithSignature, from: sender })
+
+        const { v, r, s } = signature
+        assertAmountOfEvents(receipt, 'CalldataDecoded')
+        assertEvent(receipt, 'CalldataDecoded', { expectedArgs: { data: calldata, deadline, v, r, s } })
+      })
+    })
+  })
+
+  describe('authenticate', () => {
+    let user, deadline, nonce, allowedSender, allowedFunctionality
+
+    before('mock nonces', async () => {
+      await validator.increaseNonce(externalAccount)
+      await validator.increaseNonce(sender)
+    })
+
+    const itReverts = (extraCalldata = undefined) => {
```

Comment:
Yup, but I wanted to make explicit it actually has a purpose if you pass an undefined extra calldata, do you see it useless?

Reply:
Got it, perhaps a comment would suffice :). I don't know if this communicates that well, since I'm already thinking "well `extraCalldata` would be undefined anyway by default".